File: problems/966_vowel_spellchecker/966_vowel_spellchecker.py
class Solution:
    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        def mask(word: str) -> str:
            vowels = set(["a", "e", "i", "o", "u"])
            mask_word = ""
            for c in word:
                if c in vowels:
                    mask_word += "*"
                else: 
                    mask_word += c
            return mask_word
        
        exact = set(wordlist)
        caseMap = {}
        vowelMap = {}

        for w in wordlist:
            lw = w.lower()
            caseMap.setdefault(lw, w)
            vowelMap.setdefault(mask(lw), w)
        
        res = []
        for q in queries:
            if q in exact:
                res.append(q)
            else:
                lq = q.lower()
                if lq in caseMap:
                    res.append(caseMap[lq])
                elif mask(lq) in vowelMap:
                    res.append(vowelMap[mask(lq)])
                else:
                    res.append("")
        return res

    

File: problems/966_vowel_spellchecker/966_vowel_spellchecker.java
class Solution {
    public String[] spellchecker(String[] wordlist, String[] queries) {
        Set<String> exactSet = new HashSet<>(Arrays.asList(wordlist));
        Map<String, String> caseMap = new HashMap<>();
        Map<String, String> vowelMap = new HashMap<>();

        for (String w : wordlist) {
            String lower = w.toLowerCase();
            caseMap.putIfAbsent(lower, w);
            vowelMap.putIfAbsent(mask(lower), w);
        }

        int n = queries.length;
        String[] res = new String[n];
        for (int i = 0; i < n; i++) {
            String query = queries[i];
            if (exactSet.contains(query)) {
                res[i] = query;
            } else {
                String lower = query.toLowerCase();
                if (caseMap.containsKey(lower)) {
                    res[i] = caseMap.get(lower);
                } else {
                    res[i] = vowelMap.getOrDefault(mask(lower), "");
                }
            }
        }
        return res;
    }

    private String mask(String word) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u') {
                c = '*';
            }
            sb.append(c);
        }
        return sb.toString();
    }

}